### PR TITLE
Put date skeletons behind a feature

### LINKED ISSUE #2324: Put date skeletons behind a feature
We don't currently hide the unfinished date skeletons APIs behind a feature. We should add an "experimental" feature and put them there.

### PR DESCRIPTION
**Pull request by sffc**
Fixes #2324

I added two features to icu_datetime:

- "experimental" for full-featured components bag
- "experimental_skeleton_matching" for a subset required by datagen (this is also a dependency of the "dategen" feature)

### CODE REVIEW

**Comment by robertbastian** [comment]
Activate this feature in https://github.com/unicode-org/icu4x/blob/main/components/icu/Cargo.toml#L73

**Comment by sffc** [comment]
A thorny side-issue is that this brings up questions about whether or not "datagen" is stable for clients to use. We've gone back and forth on that topic before.

File: components/datetime/src/any/datetime.rs [hunk 1]
```diff
@@ -262,10 +259,87 @@ impl DateTimeFormatter {
...
+                patterns,
+                || calendar::load_symbols_for_any_calendar_kind(data_provider, locale, kind),
+                locale_with_cal,
+                options,
+            )?,
+            calendar,
+        ))
+    }
+
+    // No docs because the "experimental" version is what docs use
```
Comment (Manishearth) [inline components/datetime/src/any/datetime.rs]: suggestion: worth noting the optional SkeletonsPattern bound in the docs of the other one ("The X bound is only needed with the experimental feature and enables blah blah")

**Review by Manishearth — DISMISSED** [review]

**Review by Manishearth — APPROVED** [review]

**Review by Manishearth — DISMISSED** [review]

**Comment by sffc** [comment]
As a side-effect, note how `provider/datagen/tests/data/work_log.wasm` got substantially smaller across this change, since it doesn't need to carry skeleton resolution logic any more. When we re-enable it, we should try to make the binary size impact as small as possible.

**Comment by sffc** [comment]
@zbraniecki The bench build is failing because the benches currently require components bag. If I do like I did in the tests where I skip over the components tests when the feature is not enabled, then the benches are going to have different results depending on what features are enabled, since enabling the experimental feature makes the benches run more examples. I could:

1. Make the benches require the "experimental" feature
2. Decide that it is fine for the benches to have different perf characteristics
3. Block this PR until I have time to split the benches into experimental and non-experimental versions

**Comment by sffc** [comment]
I think there are two issues remaining before CI is green:

1. The issue about datetime benches discussed above
2. How do we make testdata build properly when the experimental feature is disabled? We can't sprinkle `#[cfg]` everywhere because this is generated code.

**Comment by sffc** [comment]
I resolved the two issues above by:

1. Turns out there is only one bench that requires components data, "datetime/datetime_components". I simply put this one behind a `#[cfg]`.
2. I made `icu_testdata/baked` depend on `icu_datetime/experimental`. Longer term I hope we can do https://github.com/unicode-org/icu4x/issues/2383.

**Review by Manishearth — APPROVED** [review]

**Comment by sffc** [comment]
Yippee! Merged.

**Comment by zbraniecki** [comment]
> Turns out there is only one bench that requires components data, "datetime/datetime_components". I simply put this one behind a #[cfg].

That's what I would suggest. All good!

**Pull request merged**